src/Cardano/AddressDerivation/Byron.hs

```diff
@@ -0,0 +1,252 @@
+-- wallet. This increases the key length to 34 bytes, selectKey is greater than the
+-- minimum for 'generate' (32 bytes).
+--
+-- Note that our current implementation deviates from BIP-39 because we use a
+-- hash function (Blake2b) rather than key stretching with PBKDF2.
+--
+-- There are two methods of hashing the seed entropy, for different use cases.
+--
+-- 1. Normal random derivation wallet seeds. The seed entropy is hashed using
+--    Blake2b_256, inside a double CBOR serialization sandwich.
+--
+-- 2. Seeds for redeeming paper wallets. The seed entropy is hashed using
+--    Blake2b_256, without any serialization.
+hashSeed :: ScrubbedBytes -> ScrubbedBytes
+hashSeed = serialize . blake2b256 . serialize
+  where
+    serialize = BA.convert . cbor . BA.convert
+    cbor = CBOR.toStrictByteString . CBOR.encodeBytes
+
+blake2b256 :: ScrubbedBytes -> ScrubbedBytes
+blake2b256 = BA.convert . hash @ScrubbedBytes @Blake2b_256
+
+-- | Derive a symmetric key for encrypting and authenticating the address
+-- derivation path. PBKDF2 encryption using HMAC with the hash algorithm SHA512
+-- is employed.
+hdPassphrase :: XPub -> ScrubbedBytes
+hdPassphrase masterKey =
+    PBKDF2.generate
+    (PBKDF2.prfHMAC SHA512)
+    (PBKDF2.Parameters 500 32)
+    (unXPub masterKey)
+    ("address-hashing" :: ByteString)
+
+mkByronKeyFromMasterKey
+    :: XPrv
+    -> ByronKey 'RootK XPrv
+mkByronKeyFromMasterKey = unsafeMkByronKeyFromMasterKey ()
+
+unsafeMkByronKeyFromMasterKey
+    :: DerivationPath depth
+    -> XPrv
+    -> ByronKey depth XPrv
+unsafeMkByronKeyFromMasterKey derivationPath masterKey = ByronKey
+    { getKey = masterKey
+    , derivationPath
+    , payloadPassphrase = hdPassphrase (toXPub masterKey)
+    }
+
+{-------------------------------------------------------------------------------
+                                 HD derivation
+-------------------------------------------------------------------------------}
+
+-- | Derives account private key from the given root private key, using
+-- derivation scheme 1.
+deriveAccountPrivateKey
+    :: ScrubbedBytes
+    -> ByronKey 'RootK XPrv
+    -> Index 'WholeDomain 'AccountK
+    -> ByronKey 'AccountK XPrv
+deriveAccountPrivateKey pwd masterKey idx@(Index accIx) = ByronKey
+    { getKey = deriveXPrv DerivationScheme1 pwd (getKey masterKey) accIx
+    , derivationPath = idx
+    , payloadPassphrase = payloadPassphrase masterKey
+    }
+
+-- | Derives address private key from the given account private key, using
+-- derivation scheme 1.
+deriveAddressPrivateKey
+    :: ScrubbedBytes
+    -> ByronKey 'AccountK XPrv
+    -> Index 'WholeDomain 'AddressK
+    -> ByronKey 'AddressK XPrv
+deriveAddressPrivateKey pwd accountKey idx@(Index addrIx) = ByronKey
+    { getKey = deriveXPrv DerivationScheme1 pwd (getKey accountKey) addrIx
+    , derivationPath = (derivationPath accountKey, idx)
+    , payloadPassphrase = payloadPassphrase accountKey
+    }
```

Comment:
This should implement: https://input-output-hk.github.io/cardano-addresses/haddock/Cardano-AddressDerivation.html#t:HardDerivation, no?

Reply:
added in https://github.com/input-output-hk/cardano-addresses/pull/4/commits/22ce22c2de17a43e34d027ec0eb18933e086d814